contracts/zaps/ZapSteth.sol

```diff
@@ -0,0 +1,309 @@
+    /// @param _amount Amount of PT or YT to withdraw.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _slippageAllowance The slippage allowed for the internal curve stable-swap call.
+    /// @param _zeroIfPrincipal Zero if the assets to redeem are principal tokens.
+    function zapOutStEth(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _slippageAllowance,
+        int128 _zeroIfPrincipal
+    ) external {
+        _zapOut(
+            _amount,
+            _expiration,
+            _position,
+            _slippageAllowance,
+            _zeroIfPrincipal,
+            1
+        );
+    }
+
+    /// @notice There should never be any tokens in this contract.
+    /// This function can rescue any possible leftovers.
+    /// @param token The token to rescue.
+    /// @param amount The amount to rescue.
+    function rescueTokens(address token, uint256 amount) external onlyOwner {
+        if (token == address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)) {
+            amount = Math.min(address(this).balance, amount);
+            payable(msg.sender).transfer(amount);
+        } else {
+            IERC20 want = IERC20(token);
+            amount = Math.min(want.balanceOf(address(this)), amount);
+            want.transfer(msg.sender, amount);
+        }
+    }
+
+    /// @notice Helper function that takes principal or yield tokens and redeems them for ETH or stETH.
+    /// @param _amount Amount of PT or YT to withdraw.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _slippageAllowance The slippage allowed for the internal curve stable-swap call.
+    /// @param _zeroIfPrincipal Zero if the assets to redeem are principal tokens.
+    /// @param _zeroIfEth Zero if the assets to redeem for is ETH. otherwise it will be stETH.
+    function _zapOut(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _slippageAllowance,
+        int128 _zeroIfPrincipal,
+        int128 _zeroIfEth
+    ) internal {
+        ITranche tranche = _deriveTranche(address(_position), _expiration);
+
+        if (_zeroIfPrincipal == 0) {
+            tranche.transferFrom(msg.sender, address(this), _amount);
+            tranche.approve(address(tranche), _amount);
+            tranche.withdrawPrincipal(_amount, address(this));
+        } else {
+            IERC20 yt = IERC20(tranche.interestToken());
+            yt.transferFrom(msg.sender, address(this), _amount);
+            yt.approve(address(tranche), _amount);
+            tranche.withdrawInterest(_amount, address(this));
+        }
+
+        uint256 balance = want.balanceOf(address(this));
+        require(balance > 0, "no balance");
+
+        _noReentry = true;
```

Comment:
The reentrancy case isn't checked in the withdraw function just set, but I think we should prevent withdraws from being reentrantly called as well.